Bicep build timeout

**Description of the issue**

Currently when expanding bicep files, `bicep build` will timeout after 5 seconds. This works for most use cases.

Some complex deployment with multiple nested module may take longer to build, particularly if the module is stored in a private registry.

When restoring from a private registry the module is downloaded and cached to improve build times. However if the module is not cached, or a longer process such as OAuth token refresh needs to occur the build may timeout with no recourse.

**Expected behaviour**

Provide a configurable option to set the timeout of build process for the environment.

**Error output**

```text
[ERROR] Bicep (0.4.1124) compilation of 'C:\<removed>\deploy.bicep' failed with: Bicep compilation hasn't completed within the timeout window. This can be caused by errors or warnings. Check the Bicep output by running bicep build and addressing any issues.
```

**Module in use and version:**

- Module: PSRule.Rules.Azure
- Version: **1.13.2**

Captured output from `$PSVersionTable`:

```text
Name                           Value
----                           -----
PSVersion                      7.2.1
PSEdition                      Core
GitCommitId                    7.2.1
OS                             Microsoft Windows 10.0.22000
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

--- a/docs/CHANGELOG-v1.md
+++ b/docs/CHANGELOG-v1.md
@@ -14,6 +14,11 @@ See [troubleshooting guide] for a workaround to this issue.
 
 ## Unreleased
 
+What's changed since v1.13.2:
+
+- Bug fixes:
+  - Fixed bicep build timeout for complex deployments. [#1299](https://github.com/Azure/PSRule.Rules.Azure/issues/1299)
+
 ## v1.13.2
 
 What's changed since v1.13.1:

--- a/docs/setup/configuring-expansion.md
+++ b/docs/setup/configuring-expansion.md
@@ -80,6 +80,41 @@ configuration:
   AZURE_BICEP_FILE_EXPANSION: true
 ```
 
+### Bicep compilation timeout
+
+:octicons-milestone-24: v1.13.3
+
+This configuration option determines the maximum time to spend building a single Bicep source file.
+The timeout is configured in seconds.
+
+When a timeout occurs, PSRule for Azure stops the build and returns an error.
+Any resources contained within Bicep source files that exceeded the timeout are not analyzed.
+
+The default timeout is 5 seconds, however the timeout can be set to an integer between `1` and `120`.
+
+Syntax:
+
+```yaml
+configuration:
+  AZURE_BICEP_FILE_EXPANSION_TIMEOUT: int
+```
+
+Default:
+
+```yaml
+# YAML: The default AZURE_BICEP_FILE_EXPANSION_TIMEOUT configuration option
+configuration:
+  AZURE_BICEP_FILE_EXPANSION_TIMEOUT: 5
+```
+
+Example:
+
+```yaml
+# YAML: Set the AZURE_BICEP_FILE_EXPANSION_TIMEOUT configuration option to enable expansion
+configuration:
+  AZURE_BICEP_FILE_EXPANSION_TIMEOUT: 15
+```
+
 ### Require template metadata link
 
 :octicons-milestone-24: v1.7.0

--- a/docs/setup/configuring-options.md
+++ b/docs/setup/configuring-options.md
@@ -42,6 +42,9 @@ Use comments to add context.
       # Enable expansion of Azure Bicep files.
       AZURE_BICEP_FILE_EXPANSION: true
 
+      # Configure the timeout for bicep build to 15 seconds.
+      AZURE_BICEP_FILE_EXPANSION_TIMEOUT: 15
+
       # Configure the minimum AKS cluster version.
       AZURE_AKS_CLUSTER_MINIMUM_VERSION: 1.22.4
 

--- a/docs/setup/setup-bicep.md
+++ b/docs/setup/setup-bicep.md
@@ -136,6 +136,7 @@ In certain environments it may be necessary to increase the default timeout for 
 This can occur if your Bicep deployments are:
 
 - Large and complex.
+- Use nested modules.
 - Use modules restored from a registry.
 
 If you are experincing timeout errors you can increase the default timeout of 5 seconds.

--- a/docs/setup/setup-bicep.md
+++ b/docs/setup/setup-bicep.md
@@ -5,10 +5,10 @@ author: BernieWhite
 # Setup Bicep
 
 To expand Azure resources for analysis from Bicep source files the Bicep CLI is required.
-The Bicep CLI is already installed on hosted runners and agents used by GitHub Action and Azure Pipelines.
+The Bicep CLI is already installed on hosted runners and agents used by GitHub Actions and Azure Pipelines.
 
 !!! Abstract
-    This topic covers setting up support for analyzing Azure resources within Bicep source file.
+    This topic covers setting up support for analyzing Azure resources within Bicep source files.
 
 ## Installing Bicep CLI
 
@@ -35,7 +35,7 @@ For installation instructions see [Setup your Bicep development environment][2].
 
 ## Setting environment variables
 
-When expanding Bicep files, the path to the Bicep binary is required.
+When expanding Bicep files, the path to the Bicep CLI binary is required.
 By default, the `PATH` environment variable will be used to discover the binary path.
 When using this option, add the sub-directory containing the Bicep binary to the environment variable.
 
@@ -43,30 +43,26 @@ Alternatively, the path can be overridden by setting the `PSRULE_AZURE_BICEP_PAT
 When setting `PSRULE_AZURE_BICEP_PATH` specify the full path to the Bicep binary including the file name.
 File names used for Bicep binaries include `bicep`, or `bicep.exe`.
 
-For example:
+!!! Example
 
-```bash
-# Bash: Setting environment variable
-export PSRULE_AZURE_BICEP_PATH='/usr/local/bin/bicep'
-```
+    ```bash title="Bash"
+    export PSRULE_AZURE_BICEP_PATH='/usr/local/bin/bicep'
+    ```
 
-```powershell
-# PowerShell: Setting environment variable
-$Env:PSRULE_AZURE_BICEP_PATH = '/usr/local/bin/bicep';
-```
+    ```powershell title="PowerShell"
+    $Env:PSRULE_AZURE_BICEP_PATH = '/usr/local/bin/bicep';
+    ```
 
-```yaml
-# GitHub Actions: Setting environment variable
-env:
-  PSRULE_AZURE_BICEP_PATH: '/usr/local/bin/bicep'
-```
+    ```yaml title="GitHub Actions"
+    env:
+      PSRULE_AZURE_BICEP_PATH: '/usr/local/bin/bicep'
+    ```
 
-```yaml
-# Azure Pipelines: Setting environment variable
-variables:
-- name: PSRULE_AZURE_BICEP_PATH
-  value: '/usr/local/bin/bicep'
-```
+    ```yaml title="Azure Pipelines"
+    variables:
+    - name: PSRULE_AZURE_BICEP_PATH
+      value: '/usr/local/bin/bicep'
+    ```
 
 ### Using Azure CLI
 
@@ -77,30 +73,26 @@ This is explained in [Setup your Bicep development environment][3].
 
 To enable this option, set the `PSRULE_AZURE_BICEP_USE_AZURE_CLI` environment variable to `true`.
 
-For example:
+!!! Example
 
-```bash
-# Bash: Setting environment variable
-export PSRULE_AZURE_BICEP_USE_AZURE_CLI=true
-```
+    ```bash title="Bash"
+    export PSRULE_AZURE_BICEP_USE_AZURE_CLI=true
+    ```
 
-```powershell
-# PowerShell: Setting environment variable
-$Env:PSRULE_AZURE_BICEP_USE_AZURE_CLI = 'true'
-```
+    ```powershell title="PowerShell"
+    $Env:PSRULE_AZURE_BICEP_USE_AZURE_CLI = 'true'
+    ```
 
-```yaml
-# GitHub Actions: Setting environment variable
-env:
-  PSRULE_AZURE_BICEP_USE_AZURE_CLI: true
-```
+    ```yaml title="GitHub Actions"
+    env:
+      PSRULE_AZURE_BICEP_USE_AZURE_CLI: true
+    ```
 
-```yaml
-# Azure Pipelines: Setting environment variable
-variables:
-- name: PSRULE_AZURE_BICEP_USE_AZURE_CLI
-  value: true
-```
+    ```yaml title="Azure Pipelines"
+    variables:
+    - name: PSRULE_AZURE_BICEP_USE_AZURE_CLI
+      value: true
+    ```
 
   [3]: https://docs.microsoft.com/azure/azure-resource-manager/bicep/install#azure-cli
 
@@ -123,7 +115,7 @@ This option can be set within the `ps-rule.yaml` file.
 
 ```yaml
 configuration:
-  # Enable automatic expansion of bicep source files
+  # Enable automatic expansion of bicep source files.
   AZURE_BICEP_FILE_EXPANSION: true
 ```
 
@@ -135,3 +127,27 @@ configuration:
   [4]: configuring-expansion.md#bicepsourceexpansion
   [5]: ../using-bicep.md#usingparameterfiles
   [6]: ../using-templates.md#bymetadata
+
+### Configuring timeout
+
+[:octicons-book-24: Docs][7]
+
+In certain environments it may be necessary to increase the default timeout for building Bicep files.
+This can occur if your Bicep deployments are:
+
+- Large and complex.
+- Use modules restored from a registry.
+
+If you are experincing timeout errors you can increase the default timeout of 5 seconds.
+To configure the timeout, set `Configuration.AZURE_BICEP_FILE_EXPANSION_TIMEOUT` to the timeout in seconds.
+
+```yaml
+configuration:
+  # Enable automatic expansion of bicep source files.
+  AZURE_BICEP_FILE_EXPANSION: true
+
+  # Configure the timeout for bicep build to 15 seconds.
+  AZURE_BICEP_FILE_EXPANSION_TIMEOUT: 15
+```
+
+  [7]: configuring-expansion.md#bicepcompilationtimeout

--- a/docs/troubleshooting.md
+++ b/docs/troubleshooting.md
@@ -6,6 +6,27 @@ author: BernieWhite
 
 This article provides troubleshooting instructions for common errors.
 
+## Bicep compilation timeout
+
+When expanding Bicep source files you may get an error similar to the following:
+
+!!! Error
+
+    Bicep (0.4.1124) compilation of 'C:\temp\deploy.bicep' failed with: Bicep compilation hasn't completed within the timeout window. This can be caused by errors or warnings. Check the Bicep output by running bicep build and addressing any issues.
+
+This error is raised when Bicep takes longer then the timeout to build a source file.
+The default timeout is 5 seconds.
+
+You can take steps to reduce your code complexity and reduce the time a build takes by:
+
+- Removing unnecessary nested `module` calls.
+- Cache bicep modules restored from a registry in continious integration (CI) pipelines.
+
+To increase the timeout value, set the `AZURE_BICEP_FILE_EXPANSION_TIMEOUT` configuration option.
+See [Bicep compilation timeout][1] for details on how to configure this option.
+
+  [1]: setup/configuring-expansion.md#bicepcompilationtimeout
+
 ## Could not load file or assembly YamlDotNet
 
 PSRule **>=1.3.0** uses an updated version of the YamlDotNet library.

--- a/src/PSRule.Rules.Azure/Data/Bicep/BicepHelper.cs
+++ b/src/PSRule.Rules.Azure/Data/Bicep/BicepHelper.cs
@@ -20,22 +20,30 @@ namespace PSRule.Rules.Azure.Data.Bicep
 {
     internal sealed class BicepHelper
     {
+        private const int BICEP_TIMEOUT_MIN = 1;
+        private const int BICEP_TIMEOUT_MAX = 120;
+
         private static readonly char[] LINUX_PATH_ENV_SEPARATOR = new char[] { ':' };
         private static readonly char[] WINDOWS_PATH_ENV_SEPARATOR = new char[] { ';' };
 
         private readonly PipelineContext Context;
+        private readonly int _Timeout;
 
         private static BicepInfo _Bicep;
 
-        public BicepHelper(PipelineContext context)
+        public BicepHelper(PipelineContext context, int timeout)
         {
             Context = context;
+            _Timeout = timeout < BICEP_TIMEOUT_MIN ? BICEP_TIMEOUT_MIN : timeout;
+            if (_Timeout > BICEP_TIMEOUT_MAX)
+                _Timeout = BICEP_TIMEOUT_MAX;
         }
 
         internal sealed class BicepInfo
         {
             private readonly string _BinPath;
             private readonly bool _UseAzCLI;
+
             private string _Version;
 
             public BicepInfo(string binPath, bool useAzCLI)
@@ -44,7 +52,7 @@ namespace PSRule.Rules.Azure.Data.Bicep
                 _UseAzCLI = useAzCLI;
             }
 
-            internal BicepProcess Create(string sourcePath)
+            internal BicepProcess Create(string sourcePath, int timeout)
             {
                 var args = GetBicepBuildArgs(sourcePath, _UseAzCLI);
                 var startInfo = new ProcessStartInfo(_BinPath, args)
@@ -55,7 +63,7 @@ namespace PSRule.Rules.Azure.Data.Bicep
                     UseShellExecute = false,
                     WorkingDirectory = PSRuleOption.GetWorkingPath(),
                 };
-                return new BicepProcess(Process.Start(startInfo), _Version);
+                return new BicepProcess(Process.Start(startInfo), timeout, _Version);
             }
 
             internal void GetVersionInfo()
@@ -69,7 +77,7 @@ namespace PSRule.Rules.Azure.Data.Bicep
                     UseShellExecute = false,
                     WorkingDirectory = PSRuleOption.GetWorkingPath(),
                 };
-                var bicep = new BicepProcess(Process.Start(versionStartInfo));
+                var bicep = new BicepProcess(Process.Start(versionStartInfo), 5);
                 try
                 {
                     if (bicep.WaitForExit(out _))
@@ -99,17 +107,17 @@ namespace PSRule.Rules.Azure.Data.Bicep
             private readonly StringBuilder _Error;
             private readonly AutoResetEvent _ErrorWait;
             private readonly AutoResetEvent _OutputWait;
+            private readonly int _Interval;
             private readonly int _Timeout;
-            private readonly int _Retry;
 
             private bool _Disposed;
 
-            public BicepProcess(Process process, string version = null)
+            public BicepProcess(Process process, int timeout, string version = null)
             {
                 _Output = new StringBuilder();
                 _Error = new StringBuilder();
-                _Timeout = 1000;
-                _Retry = 5;
+                _Interval = 1000;
+                _Timeout = timeout;
 
                 Version = version;
                 _Process = process;
@@ -132,12 +140,12 @@ namespace PSRule.Rules.Azure.Data.Bicep
                 if (!_Process.HasExited)
                 {
                     var timeoutCount = 0;
-                    while (!_Process.WaitForExit(_Timeout) && !_Process.HasExited && timeoutCount < _Retry)
+                    while (!_Process.WaitForExit(_Interval) && !_Process.HasExited && timeoutCount < _Timeout)
                         timeoutCount++;
                 }
 
                 exitCode = _Process.HasExited ? _Process.ExitCode : -1;
-                return _Process.HasExited && _ErrorWait.WaitOne(_Timeout) && _OutputWait.WaitOne();
+                return _Process.HasExited && _ErrorWait.WaitOne(_Interval) && _OutputWait.WaitOne();
             }
 
             public string GetOutput()
@@ -264,9 +272,9 @@ namespace PSRule.Rules.Azure.Data.Bicep
             return results.ToArray();
         }
 
-        private static JObject ReadBicepFile(string path)
+        private JObject ReadBicepFile(string path)
         {
-            var bicep = GetBicep(path);
+            var bicep = GetBicep(path, _Timeout);
             if (bicep == null)
                 throw new BicepCompileException(string.Format(Thread.CurrentThread.CurrentCulture, PSRuleResources.BicepNotFound), null, path, null);
 
@@ -305,12 +313,12 @@ namespace PSRule.Rules.Azure.Data.Bicep
             }
         }
 
-        private static BicepProcess GetBicep(string sourcePath)
+        private static BicepProcess GetBicep(string sourcePath, int timeout)
         {
             if (_Bicep == null)
                 _Bicep = GetBicepInfo();
 
-            return _Bicep?.Create(sourcePath);
+            return _Bicep?.Create(sourcePath, timeout);
         }
 
         private static BicepInfo GetBicepInfo()

--- a/src/PSRule.Rules.Azure/Runtime/Helper.cs
+++ b/src/PSRule.Rules.Azure/Runtime/Helper.cs
@@ -27,7 +27,7 @@ namespace PSRule.Rules.Azure.Runtime
             return ExpressionParser.IsExpression(expression);
         }
 
-        public static PSObject[] GetResources(string parameterFile)
+        public static PSObject[] GetResources(string parameterFile, int timeout)
         {
             var context = GetContext();
             var linkHelper = new TemplateLinkHelper(context, PSRuleOption.GetWorkingPath(), true);
@@ -36,13 +36,13 @@ namespace PSRule.Rules.Azure.Runtime
                 return null;
 
             return IsBicep(link.TemplateFile) ?
-                GetBicepResources(link.TemplateFile, link.ParameterFile, null) :
+                GetBicepResources(link.TemplateFile, link.ParameterFile, null, timeout) :
                 GetTemplateResources(link.TemplateFile, link.ParameterFile, context);
         }
 
-        public static PSObject[] GetBicepResources(string bicepFile, PSCmdlet commandRuntime)
+        public static PSObject[] GetBicepResources(string bicepFile, PSCmdlet commandRuntime, int timeout)
         {
-            return GetBicepResources(bicepFile, null, commandRuntime);
+            return GetBicepResources(bicepFile, null, commandRuntime, timeout);
         }
 
         public static string GetMetadataLinkPath(string parameterFile, string templateFile)
@@ -79,11 +79,12 @@ namespace PSRule.Rules.Azure.Runtime
             return Path.GetExtension(path) == ".bicep";
         }
 
-        private static PSObject[] GetBicepResources(string templateFile, string parameterFile, PSCmdlet commandRuntime)
+        private static PSObject[] GetBicepResources(string templateFile, string parameterFile, PSCmdlet commandRuntime, int timeout)
         {
             var context = GetContext(commandRuntime);
             var bicep = new BicepHelper(
-                context
+                context,
+                timeout
             );
             return bicep.ProcessFile(templateFile, parameterFile);
         }

--- a/src/PSRule.Rules.Azure/rules/Conventions.Rule.ps1
+++ b/src/PSRule.Rules.Azure/rules/Conventions.Rule.ps1
@@ -19,8 +19,9 @@ Export-PSRuleConvention 'Azure.ExpandTemplate' -If { $Configuration.AZURE_PARAME
     "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json`#"
 ), $True) } -Begin {
     Write-Verbose "[Azure.ExpandTemplate] -- Expanding parameter file: $($TargetObject.FullName)";
+    $timeout = $Configuration.GetIntegerOrDefault('AZURE_BICEP_FILE_EXPANSION_TIMEOUT', 5);
     try {
-        $data = [PSRule.Rules.Azure.Runtime.Helper]::GetResources($TargetObject.FullName);
+        $data = [PSRule.Rules.Azure.Runtime.Helper]::GetResources($TargetObject.FullName, $timeout);
         if ($Null -ne $data) {
             $PSRule.Import($data);
         }
@@ -61,8 +62,9 @@ Export-PSRuleConvention 'Azure.BicepInstall' -If { $Configuration.AZURE_BICEP_FI
 
 Export-PSRuleConvention 'Azure.ExpandBicep' -If { $Configuration.AZURE_BICEP_FILE_EXPANSION -eq $True -and $TargetObject.Extension -eq '.bicep' } -Begin {
     Write-Verbose "[Azure.ExpandBicep] -- Expanding bicep source: $($TargetObject.FullName)";
+    $timeout = $Configuration.GetIntegerOrDefault('AZURE_BICEP_FILE_EXPANSION_TIMEOUT', 5);
     try {
-        $data = [PSRule.Rules.Azure.Runtime.Helper]::GetBicepResources($TargetObject.FullName, $PSCmdlet);
+        $data = [PSRule.Rules.Azure.Runtime.Helper]::GetBicepResources($TargetObject.FullName, $PSCmdlet, $timeout);
         if ($Null -ne $data) {
             $PSRule.Import($data);
         }

